### PR TITLE
[14.0][MIG] sale_order_secondary_unit

### PR DESCRIPTION
**Pull request by Kev-Roche**
@sergio-teruel @carlosdauden @alexeirivera87 @hparfr
Replace inactive migration OCA#1428 and add improvements from OCA/product-attribute#848

### CODE REVIEW

**Comment by Kev-Roche** [comment]
Fail because of  OCA/product-attribute#848 dependency, migration in progress.

**Comment by pedrobaeza** [comment]
Dependency is now merged. Can you rebase/relaunch?

**Comment by sergio-teruel** [comment]
Travis is red!

**Comment by Kev-Roche** [comment]
@sergio-teruel indeed, It's coming from the product_uom_qty which is now computed like in stock_secondary_unit https://github.com/OCA/stock-logistics-warehouse/pull/1053/files (in order to use the function in the mixin model OCA/product-attribute#848).
But here, the module sale_order_qty_change_no_recompute is not allowing the product_uom_qty as computed. Do you have an idea of the best solution?

**Comment by Bilbonet** [comment]
in the sale order line if you put 0 in quantity

Traceback (most recent call last):
  File "/home/bilbonet/odoo14/server/odoo/http.py", line 640, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/bilbonet/odoo14/server/odoo/http.py", line 316, in _handle_exception
    raise exception.with_traceback(None) from new_cause
ZeroDivisionError: float division by zero

**Comment by Kev-Roche** [comment]
> ZeroDivisionError: float division by zero

Hi @Bilbonet, thank you for highlighting this error, I fixed it in the last commit, also here https://github.com/OCA/product-attribute/pull/902 to make it work.

**Comment by Bilbonet** [comment]
Hi @Kev-Roche 
Now, I have this error if I put 0 in qty or in secondary qty. 

Traceback (most recent call last):
  File "/home/bilbonet/odoo14/server/odoo/http.py", line 640, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/bilbonet/odoo14/server/odoo/http.py", line 316, in _handle_exception
    raise exception.with_traceback(None) from new_cause
ValueError: Compute method failed to assign sale.order.line(<NewId origin=11>,).secondary_uom_unit_price

**Comment by Kev-Roche** [comment]
> Hi @Kev-Roche
> Now, I have this error if I put 0 in qty or in secondary qty.
> 
> Traceback (most recent call last):
> File "/home/bilbonet/odoo14/server/odoo/http.py", line 640, in _handle_exception
> return super(JsonRequest, self)._handle_exception(exception)
> File "/home/bilbonet/odoo14/server/odoo/http.py", line 316, in _handle_exception
> raise exception.with_traceback(None) from new_cause
> ValueError: Compute method failed to assign sale.order.line(,).secondary_uom_unit_price

My bad, I misunderstood the issue, it will be ok now. No need to change OCA/product-attribute#902 anymore

**Review by Bilbonet — APPROVED** [review]
I've chequed the module and it works fine for me 
Thanks!! :-)

File: sale_order_secondary_unit/models/sale_order.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        for line in self:
+            if line.secondary_uom_id:
+                try:
+                    line.secondary_uom_unit_price = (
+                        line.price_subtotal / line.secondary_uom_qty
+                    )
+                except ZeroDivisionError:
+                    line.secondary_uom_unit_price = 0
+            else:
+                line.secondary_uom_unit_price = 0
```
Comment (pedrobaeza) [inline sale_order_secondary_unit/models/sale_order.py:67]:
```suggestion
            if line.secondary_uom_id and line.secondary_uom_qty:
                line.secondary_uom_unit_price = (
                    line.price_subtotal / line.secondary_uom_qty
                )
            else:
                line.secondary_uom_unit_price = 0
```

**Review by pedrobaeza — COMMENTED** [review]
Travis is red

File: sale_order_secondary_unit/models/sale_order.py (hunk 1 above)
Comment (Kev-Roche) [reply sale_order_secondary_unit/models/sale_order.py:67]: Travis was already red before, the module sale_order_qty_change_no_recompute is in conflict with the computed product_uom_qty field in this module.

Comment (luffah) [reply sale_order_secondary_unit/models/sale_order.py:67]: As not being a Travis erudite, it looks strange to me.

I ran test cases 1 by 1 in a Python shell, (with `setupClass()` between) and re-run tests with `odoo --test-enable` and no error is raised

Too, i tested with odoo standard --test-enable both modules installed both pass without error (i had to change a partner reference for sale_order_qty_change_no_recompute test setup).

_Note_
_In order to install both modules :  `git merge OCA/14.0` in my local copy of akretion branch_

**Comment by francesco-ooops** [comment]
@Kev-Roche , any plan to finish this? Otherwise we can pick it up

**Review by hparfr — APPROVED** [review]

**Comment by Kev-Roche** [comment]
@francesco-ooops, It seems that the problem is a conflict with sale_order_qty_change_no_recompute module. No idea what is the cleanest way to do it.

**Comment by carlosdauden** [comment]
ping @sergio-teruel

**Comment by sergio-teruel** [comment]
@Kev-Roche Could you change the implementation of this module using the product_secondary_unit mixin??

**Review by victoralmau — APPROVED** [review]

**Comment by victoralmau** [comment]
Please, cherry-pick https://github.com/OCA/sale-workflow/pull/2002#event-6541248480

**Comment by Bilbonet** [comment]
Hi @Kev-Roche 
It could be cool if you could implement changes suggested.
Thanks!

**Comment by sergio-teruel** [comment]
@Kev-Roche Can you add this commit https://github.com/OCA/sale-workflow/pull/2237
Thanks!!

**Comment by Kev-Roche** [comment]
> @Kev-Roche Can you add this commit #2237 Thanks!!

Done

**Pull request merged**